_posts/2023-11-17-android-platforms.md

```diff
@@ -0,0 +1,115 @@
+```
+
+Every Android `platform` should use the [`@platforms//os:android`](https://github.com/bazelbuild/platforms/blob/33a3b209f94856193266871b1545054afb90bb28/os/BUILD#L36) OS constraint. To migrate the CPU constraint, check this chart:
+
+CPU Value     | Platform
+------------- | ------------------------------------------
+`armeabi-v7a` | `@platforms//cpu:arm`
+`arm64-v8a`   | `@platforms//cpu:arm64`
+`x86`         | `@platforms//cpu:x86_32`
+`x86_64`      | `@platforms//cpu:x86_64`
+
+And, of course, for a multi-architecture APK, you pass multiple labels, for
+example: `--android_platforms=//:arm64,//:x86_64` (assuming you defined those in
+your top-level `BUILD.bazel` file).
+
+### Are any other flags changing?
+
+Yes, as part of this several legacy flags have been deprecated and are no-ops.
+
+Deprecated flags (and their new equivalents):
+
+- `--android_cpu`: Like `--fat_apk_cpu`, this is replaced by
+  `--android_platforms`
+- `--android_sdk`: Android SDKs are now registered toolchains, so if you have
+  multiple `android_sdk_repository` repository rules defined, you can use
+  `register_toolchains` and `--extra_toolchains` to change the order they are
+  used in.
+- `--android_crosstool_top`: Similarly to the Android SDK, the
+  `android_ndk_repository` rule exposes the NDK as a C++ toolchain, using the
+  standard toolchain resolution. You can use `register_toolchains` and
+  `--extra_toolchains` to change the selection order for these.
+
+### Can I test this out?
+
+Yes, download the latest [Bazel 7.0 release
+candidate](https://releases.bazel.build/7.0.0/rc3/index.html) (or use
+[Bazelisk](https://github.com/bazelbuild/bazelisk) with
+`USE_BAZEL_VERSION=last_rc`) and update your build flags. Please [file an
+issue](https://github.com/bazelbuild/bazel/issues/new/choose) if you encounter
+any problems.
+
+### Do I have to migrate?
+
+The legacy functionality is still present in the Bazel 7.0 LTS branch. You can
+disable the changes by passing the flag `--noincompatible_enable_android_toolchain_resolution`.
+
+If you find any issues with the new functionality, please [file an
+issue](https://github.com/bazelbuild/bazel/issues/new/choose) so that we can
+address it.
+
+However, note that the legacy functionality will be deleted in short order and
+will not be present in Bazel 8.0 when that is released.
+
+## Why are things changing?
+
+As part of the overall [platforms and
+toolchains](https://bazel.build/extending/platforms) project, all of Bazel is
+moving away from using `--cpu` flags to express target platforms, and towards
+the richer variety allowed by the `platform` rule. The current usage of `--cpu`
+is not actually as a single CPU architecture: we currently capture different
+things like "Linux on 64-bit intel" (`--cpu=k8`), "macOS on 64-bit-intel"
+(`--cpu=darwin_x86_64`), "Windows on ARM" (`--cpu=arm64_windows`): and that's
+just a few common OSes.
+
+Once we also want to describe Android and iOS devices, or custom development
```

Comment:
"Once we also want to ..." seems a little awkward.

Reply:
Done.